assembly/src/assembler/context.rs

```diff
@@ -1,254 +1,131 @@
-use super::{
-    AssemblyError, CallSet, CodeBlock, CodeBlockTable, Kernel, LibraryPath, NamedProcedure,
-    Procedure, ProcedureCache, ProcedureId, ProcedureName, RpoDigest,
-};
+use alloc::{boxed::Box, sync::Arc};
+
+use super::{procedure::CallSet, ArtifactKind, GlobalProcedureIndex, Procedure};
 use crate::{
-    ast::{ModuleAst, ProgramAst},
-    utils::{collections::*, string::*},
+    ast::{FullyQualifiedProcedureName, Visibility},
+    diagnostics::SourceFile,
+    AssemblyError, LibraryPath, RpoDigest, SourceSpan, Span, Spanned,
 };
+use vm_core::code_blocks::CodeBlock;
 
-// ASSEMBLY CONTEXT
-// ================================================================================================
-
-/// Contains information about compilation of a program or a kernel module.
+/// An [AssemblyContext] is used to store configuration and state pertaining to the current
+/// compilation of a module/procedure by an [crate::Assembler].
 ///
-/// Assembly context contains a stack of [ModuleContext]'s, each of which, in turn, contains a
-/// stack of [ProcedureContext]'s. Thus, at any point in time, we are in a context of compiling a
-/// procedure within a module, and we have access to the info about the current module/procedure
-/// tuple bing compiled.
+/// The context specifies context-specific configuration, the type of artifact being compiled,
+/// the current module being compiled, and the current procedure being compiled.
+///
+/// To provide a custom context, you must compile by invoking the
+/// [crate::Assembler::compile_in_context] API, which will use the provided context in place of
+/// the default one generated internally by the other `compile`-like APIs.
+#[derive(Default)]
 pub struct AssemblyContext {
-    module_stack: Vec<ModuleContext>,
-    is_kernel: bool,
-    kernel: Option<Kernel>,
+    /// What kind of artifact are we assembling
+    kind: ArtifactKind,
+    /// When true, this permits calls to refer to procedures which are not locally available,
+    /// as long as they are referenced by MAST root, and not by name. As long as the MAST for those
+    /// roots is present when the code is executed, this works fine. However, if the VM tries to
+    /// execute a program with such calls, and the MAST is not available, the program will trap.
```

Comment:
I assumed we handled it somewhat sanely rather than panicking, hence my description of the behavior as a "trap", which refers to anything that causes execution to halt unexpectedly, but in a controlled manner (i.e. the VM doesn't just segfault or something wild like that). If that's not the case today, it probably should be, since I anticipate this scenario could arise frequently in practice.

Reply:
I think we just return an error, i.e. [the VM panics](https://github.com/0xPolygonMiden/miden-vm/blob/322a70bc2f4a4e52c3a4a0d4990d01327fc41c54/processor/src/lib.rs#L344-L346)